### ISSUE TITLE
CA1508 (Avoid dead conditional code) - false positive with constructor call in try block

### ISSUE THREAD

**Issue by kindermannhubert** [dotnet/roslyn-analyzers#5002]
### Analyzer

**Diagnostic ID**: [CA1508](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/CA1508): `Avoid dead conditional code`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 5.0.3 (Latest)

### Describe the bug

False positive in following code:
```c#
static void M()
{
    C c = null;
    try
    {
        c = new C();
    }
    finally
    {
        c?.ToString(); //CA1508: 'c' is never 'null'.
    }
}

class C
{
    public C() { throw new(); }
}
```

'c' will be 'null' in finally block.
Moreover when '#nullable enable' is active, compiler correctly warns you that 'c' can be null, when you use only 'c.ToString()'.

### Expected behavior
Diagnostic is not reported in this case (independently on ctor implementation).

**Comment by kindermannhubert** [comment]
Following example is similar but I'm not sure if it's exactly same problem in analyzer. Here's another false positive:
```c#
static void M1(object notNull)
{
    if (notNull is null) return;
            
    object o = null;
    try
    {
        M2(); //could throw
        o = notNull;
    }
    finally
    {
        o?.ToString(); //CA1508: 'o' is never 'null'.
    }
}
static void M2() { /*...*/ }
```

**Comment by Thieum** [comment]
Same goes with the recommended solution for avoiding double dispose:

```
            FileStream fs = null;
            try
            {
                fs = new FileStream(filename, FileMode.Open);
                using var r = new BinaryReader(fs);
                fs = null;
                r.Read(data32Master, 0, bmp.Size.Width * bmp.Size.Height * 4);
            }
            finally
            {
                if (fs != null) // CA1508 FP
                {
                    fs.Dispose();
                }
            }
```

**Comment by mavasani** [comment]
Duplicate of https://github.com/dotnet/roslyn-analyzers/issues/5002
